SIPS/sip-20.md

```diff
@@ -0,0 +1,155 @@
+---
+sip: 20
+title: External data (WebSockets and other custom data or events)
+status: Draft
+author: David Drazic (@david0xd)
+created: 2023-12-15
+---
+
+## Abstract
+
+This SIP proposes changes to the Snap manifest and additional request handler that allows Snap developers to implement
+functionality that is invoked when external data events occur. This proposal outlines some of the details around this feature.
+
+The purpose of the initial version of this SIP is to provide support for WebSockets.
+This SIP only proposes one-way communication.
+This SIP is made with extensibility in mind and can be updated to handle more external data sources (like blockchain
+events, etc.).
+
+## Motivation
+
+Snaps are currently isolated from outside events and rely on combination of user actions or CRON jobs and fetch
+endowments in order to receive extra information from independent remote sources distributed over the internet.
+
+This proposal provides an improvement that enables one way communication between a Snap and an external data or event
+source.
+The idea is that the external event source sends arbitrary data to the specific controller with listener running
+inside a wallet. The controller then sends the data to the Snap which runs execution of this feature specific RPC
+request handler.
+
+## Specification
+
+> Formal specifications are written in Typescript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "SHOULD", "RECOMMENDED" and "MAY" written in uppercase in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Manifest
+
+This SIP specifies a permission named `endowment:external-data`.
+
+The new field can be specified as follows in a `snap.manifest.json` file:
+
+```json
+{
+  "initialPermissions": {
+    "endowment:external-data": [
+      {
+        "type": "websocket",
+        "url": "wss://example.com/binary/endpoint",
+        "dataType": "binary"
+      },
+      {
+        "type": "websocket",
+        "url": "wss://example.com/text/endpoint",
+        "dataType": "text"
+      },
+      {
+        "type": "blockchain",
+        "chainId": "eip155:1",
+        "events": [
+          "block"
+        ]
+      }
+    ]
+  }
+}
+```
+
+### Permission caveats
+
+Each caveat object MUST have `type` property that SHOULD be `websocket` or `blockchain`.
```

Comment:
We should limit this SIP to just `websocket`.

Reply:
Updated. Only websocket left.